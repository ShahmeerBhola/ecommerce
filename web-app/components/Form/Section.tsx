import React from 'react';

import Field from './Field';
import Text from '../Text';

type Props<IV> = {
    fields: Forms.Field<IV>[];
    hidden: boolean;
    label?: string;
    yMargin?: string;
    noXPadding?: boolean;
    children?: React.ReactNode;
    onChangeHandlers?: Forms.ChangeHandlers<IV>;
    onBlurHandlers?: Forms.ChangeHandlers<IV>;
    fieldOptions?: Forms.FieldOptions<IV>;
    disabledFields?: Forms.DisabledFields<IV>;
    overriddenFields?: Forms.OverriddenFields<IV>;
};

const FormSection = <IV,>({
    children,
    fields,
    label,
    yMargin,
    noXPadding,
    hidden = false,
    onChangeHandlers,
    onBlurHandlers,
    fieldOptions,
    disabledFields,
    overriddenFields,
}: Props<IV>): JSX.Element | null => {
    let yMarg = 'my-3';
    if (yMargin === '') {
        yMarg = '';
    } else if (yMargin) {
        yMarg = yMargin;
    }

    return hidden ? null : (
        <div className={`flex flex-wrap ${yMarg}`}>
            {!!label && (
                <Text color="text-gray-700" className="w-full mb-3 font-medium" size="text-lg">
                    {label}
                </Text>
            )}
            {fields.map(({ name: fieldName, type, placeholder, className, ...rest }) => {
                const name = fieldName as keyof IV;
                return (
                    <Field<IV>
                        key={name as string}
                        name={name as string}
                        type={type}
                        placeholder={placeholder}
                        className={className}
                        noXPadding={noXPadding}
                        onChange={onChangeHandlers && onChangeHandlers[name]}
                        onBlur={onBlurHandlers && onBlurHandlers[name]}
                        fieldOptions={fieldOptions && fieldOptions[name]}
                        disabled={disabledFields && disabledFields[name]}
                        overrideContent={overriddenFields && overriddenFields[name]}
                        {...rest}
                    />
                );
            })}
            {children}
        </div>
    );
};

export default FormSection;
