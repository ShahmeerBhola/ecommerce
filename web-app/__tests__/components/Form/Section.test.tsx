const Field: React.FunctionComponent = () => (
  <div>
    <p>Mocked field</p>
    <input type="text" />
  </div>
);

jest.mock('../../../components/Form/Field', () => ({
  __esModule: true,
  default: Field,
}));

import React from 'react';
import renderer from 'react-test-renderer';
import Section from '../../../components/Form/Section';

describe('Section component', () => {
  type IV = {
    first_name: '';
  };

  test('it renders properly when hidden is false', () => {
    const tree = renderer
      .create(
        <Section<IV>
          fields={[
            {
              name: 'first_name',
              type: 'text',
              placeholder: 'hello',
            },
          ]}
          hidden={false}
        >
          <div>
            <p>Yooo</p>
          </div>
        </Section>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly when hidden is true', () => {
    const tree = renderer
      .create(
        <Section<IV>
          fields={[
            {
              name: 'first_name',
              type: 'text',
              placeholder: 'hello',
            },
          ]}
          hidden
        >
          <div>
            <p>Yooo</p>
          </div>
        </Section>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly when label prop is provided', () => {
    const tree = renderer
      .create(
        <Section<IV>
          fields={[
            {
              name: 'first_name',
              type: 'text',
              placeholder: 'hello',
            },
          ]}
          hidden={false}
          label="My Section Label"
        >
          <div>
            <p>Yooo</p>
          </div>
        </Section>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
