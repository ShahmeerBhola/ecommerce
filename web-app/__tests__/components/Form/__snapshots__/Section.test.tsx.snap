// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section component it renders properly when hidden is false 1`] = `
<div
  className="flex flex-wrap my-3"
>
  <div>
    <p>
      Mocked field
    </p>
    <input
      type="text"
    />
  </div>
  <div>
    <p>
      Yooo
    </p>
  </div>
</div>
`;

exports[`Section component it renders properly when hidden is true 1`] = `null`;

exports[`Section component it renders properly when label prop is provided 1`] = `
<div
  className="flex flex-wrap my-3"
>
  <p
    className="font-secondary font-normal text-gray-700 text-lg w-full mb-3 font-medium"
  >
    My Section Label
  </p>
  <div>
    <p>
      Mocked field
    </p>
    <input
      type="text"
    />
  </div>
  <div>
    <p>
      Yooo
    </p>
  </div>
</div>
`;
